src/utils/collectStyledComponentData.js

```diff
@@ -27,127 +48,244 @@ const getAttrsType = (node) => {
-          ? node.tag.arguments[0].name
-          : node.tag.callee.object.arguments[0].name;
-
-        /**
-         * enable checking custom components
-         * @see https://github.com/jsx-eslint/eslint-plugin-jsx-a11y#usage
-         */
-        const componentMap = context.settings?.['jsx-a11y']?.components ?? {};
-
-        // styled(StyledComponent)`` || styled(StyledComponent).attrs(...)``
-        if (styledComponentsDict[ancestorScName]) {
-          ({ attrs } = styledComponentsDict[ancestorScName]);
-          ({ tag } = styledComponentsDict[ancestorScName]);
+        if (!styledComponentName) return;
+
+        // styled.?({ ... })
+        if (node.callee?.type === 'MemberExpression' && isStyledIdentifier(node.callee.object)) {
+          // styled.div({ ... })
+          if (node.callee.property?.type === 'Identifier' && node.callee.property.name) {
+            tag = node.callee.property.name;
+
+            if (!tag) return;
+
+            styledComponentsDict[styledComponentName] = {
+              name: styledComponentName,
+              attrs: [],
+              tag: tag,
+            };
+          }
         }
 
-        // styled(CustomComponent)`` || styled(CustomComponent).attrs(...)``
-        if (componentMap[ancestorScName]) {
-          tag = componentMap[ancestorScName];
+        // styled(...)(...)
+        if (node.callee?.type === 'CallExpression' && isStyledIdentifier(node.callee.callee)) {
+          let arg = node.callee.arguments[0];
+
+          if (!arg) return;
+
+          // styled('div')(...)
+          if (arg.type === 'Literal') {
+            tag = arg.value;
+
+            if (!tag) return;
+
+            styledComponentsDict[styledComponentName] = {
+              name: styledComponentName,
+              attrs: [],
+              tag: tag,
+            };
+          }
+
+          // TODO: Consider supporting templates like styled(`div`)(...)
+          if (arg.type !== 'Identifier') return;
+
+          // styled(StyledComponent)({ ... })
+
+          let attrs = [];
+          let ancestorScName = arg.name;
+
+          if (styledComponentsDict[ancestorScName]) {
+            // Add attrs if the ancestor has them
+            attrs = styledComponentsDict[ancestorScName].attrs;
+            tag = styledComponentsDict[ancestorScName].tag;
+          }
+
+          // styled(CustomComponent)({ ...})
+          if (componentMap[ancestorScName]) {
+            tag = componentMap[ancestorScName];
+          }
+
+          if (!tag) return;
+
+          styledComponentsDict[styledComponentName] = {
+            name: styledComponentName,
+            attrs: attrs,
+            tag: tag,
+          };
         }
+      } catch (error) {
+        context.report({
+          message: 'Unable to parse styled component: {{ message }}',
+          node,
+          data: { message: error.message, stack: error.stack },
+        });
       }
+    },
+    TaggedTemplateExpression(node) {
```

Comment:
I've edited the indent by making the parent have a block body but otherwise tried to leave `TaggedTemplateExpression` alone for this MR